shenyu-client/shenyu-client-grpc/src/main/java/org/apache/shenyu/client/grpc/GrpcClientEventListener.java

```diff
@@ -34,165 +34,136 @@
 
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
-import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Objects;
-import java.util.Optional;
-import java.util.Properties;
 import java.util.stream.Collectors;
 
 /**
  * The type Shenyu grpc client event listener.
  */
-public class GrpcClientEventListener implements ApplicationListener<ContextRefreshedEvent> {
-
-    private static final Logger LOG = LoggerFactory.getLogger(GrpcClientEventListener.class);
-
-    private final ShenyuClientRegisterEventPublisher publisher = ShenyuClientRegisterEventPublisher.getInstance();
-
-    private final String contextPath;
-
-    private final String ipAndPort;
-
-    private final String host;
-
-    private final int port;
-
+public class GrpcClientEventListener extends AbstractContextRefreshedEventListener<BindableService, ShenyuGrpcClient> {
+    
     private final List<ServerServiceDefinition> serviceDefinitions = Lists.newArrayList();
-
+    
     /**
      * Instantiates a new Shenyu client bean post processor.
      *
      * @param clientConfig the shenyu grpc client config
      * @param shenyuClientRegisterRepository the shenyuClientRegisterRepository
      */
     public GrpcClientEventListener(final PropertiesConfig clientConfig, final ShenyuClientRegisterRepository shenyuClientRegisterRepository) {
-        Properties props = clientConfig.getProps();
-        String contextPath = props.getProperty(ShenyuClientConstants.CONTEXT_PATH);
-        String ipAndPort = props.getProperty(ShenyuClientConstants.IP_PORT);
-        String port = props.getProperty(ShenyuClientConstants.PORT);
-        if (StringUtils.isAnyBlank(contextPath, ipAndPort, port)) {
+        super(clientConfig, shenyuClientRegisterRepository);
+        if (StringUtils.isAnyBlank(getContextPath(), getIpAndPort(), getPort())) {
             throw new ShenyuClientIllegalArgumentException("grpc client must config the contextPath, ipAndPort");
         }
-        this.ipAndPort = ipAndPort;
-        this.contextPath = contextPath;
-        this.host = props.getProperty(ShenyuClientConstants.HOST);
-        this.port = Integer.parseInt(port);
-        publisher.start(shenyuClientRegisterRepository);
     }
-
+    
     @Override
-    public void onApplicationEvent(@NonNull final ContextRefreshedEvent contextRefreshedEvent) {
-        Map<String, BindableService> beans = contextRefreshedEvent.getApplicationContext().getBeansOfType(BindableService.class);
-        for (Map.Entry<String, BindableService> entry : beans.entrySet()) {
+    protected Map<String, BindableService> getBeans(final ApplicationContext context) {
+        Map<String, BindableService> result = context.getBeansOfType(BindableService.class);
+        for (Entry<String, BindableService> entry : result.entrySet()) {
             exportJsonGenericService(entry.getValue());
```

Comment:
`exportJsonGenericService(entry.getValue());` should be put to `#handle`.

Reply:
done.